src/compiler/program.ts

```diff
@@ -211,6 +221,77 @@ namespace ts {
             return classifiableNames;
         }
 
+        function tryReuseStructureFromOldProgram(): boolean {
+            if (!host.hasChanges) {
+                // host does not support method 'hasChanges'
+                return false;
+            }
+            if (!oldProgram) {
+                return false;
+            }
+            
+            Debug.assert(!oldProgram.structureIsReused);
+
+            // there is an old program, check if we can reuse its structure
+            let oldRootNames = oldProgram.getRootFileNames();
+            if (rootNames.length !== oldRootNames.length) {
+                // different amount of root names - structure cannot be reused
+                return false;
+            }
+            
+            for (let i = 0; i < rootNames.length; i++) {
+                if (oldRootNames[i] !== rootNames[i]) {
+                    // different order of root names - structure cannot be reused
+                    return false;
+                }
+            }
+            
+            // check if program source files has changed in the way that can affect structure of the program
+            let newSourceFiles: SourceFile[] = [];
+            for (let oldSourceFile of oldProgram.getSourceFiles()) {
+                let newSourceFile: SourceFile;
+                if (host.hasChanges(oldSourceFile)) {

```

Comment:
technically there is difference, in LS scenarios querying for version can be slightly less expensive because we this is just a lookup in host cache where is getting new file involves doing more stuff